OSDOCS-10396:ANP audit logging (#75310)
openshift/openshift-docs: modules/nw-anp-audit-logging-concept.adoc

.Example of annotation for `AdminNetworkPolicy` CR
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  annotations:
    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "alert", "pass" : "warning" }'
  name: anp-tenant-log
spec:
  priority: 5
  subject:
    namespaces:
      matchLabels:
        tenant: red
  ingress:
    - name: "allow-all-ingress-tenant-1"
      action: "Allow"
      from:
      - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              tenant: blue
        podSelector:
          matchLabels:
            custom-anp: tenant-1
  egress:
    - name: "deny-ns-tenant-1"
      action: "Deny"
      to:
      - namespaces:
           matchLabels:
             tenant: blue
    - name: "allow-ns-tenant-2"
      action: "Allow"
      to:
      - namespaces:
           matchLabels:
             tenant: orange
    - name: "pass-ns-tenant-3"
      action: "Pass"
      to:
      - namespaces:
           matchLabels:
             tenant: pink
----
====

> Logs are generated whenever a specific OVN ACL triggers an event that meets the action criteria set in your logging annotation. In the example, an event in which the `tenant: red` is denied egress to the `tenant: orange` a log is generated.

Reviewer: same comment here: could you give a small preface like "tenant:blue" is set of all namespaces with the label "tenant:blue" ? same for red
so its about defining what we meant by tenant here as a set of colored namespaces, so blue tenant is a set of namespaces with the label tenant:blue and red tenant is the set of namespaces with the label "red"

> Logs are generated whenever a specific OVN ACL triggers an event that meets the action criteria set in your logging annotation. In the example, an event in which all of the namespaces with the label `tenant: red` are denied egress to all of the namespaces with the label `tenant: orange`, a log is generated.

Reviewer: I see this has come based on https://github.com/openshift/openshift-docs/pull/75310#discussion_r1627650553 but remember that "a tenant red is set of all namespaces with label "tenant:red". that's the preface. But when you write this sentence as its now ^ it becomes wrong. So an event where "any of the namespace with the label "tenant:red" is denied egress to any of the namespace with the label "tenant:orange" a log is generated is the right way. Does that make sense from technical understanding pov? I think this ties in to your confusion of "labels" concept in K8s. Perhaps https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/ will help

> Logs are generated whenever a specific OVN ACL triggers an event that meets the action criteria set in your logging annotation. In the example, an event in which all of the namespaces with the label `tenant: red` are denied egress to all of the namespaces with the label `tenant: orange`, a log is generated.

Reviewer: ```suggestion
Logs are generated whenever a specific OVN ACL triggers an event that meets the action criteria set in your logging annotation. In the example, an event in which any of the namespaces with the label `tenant: red` is denied egress to any of the namespaces with the label `tenant: orange`, a log is generated.
```

> Logs are generated whenever a specific OVN ACL triggers an event that meets the action criteria set in your logging annotation. In the example, an event in which any of the namespaces with the label `tenant: red` are denied egress to any of the namespaces with the label `tenant: orange`, a log is generated.

Reviewer 2: ```suggestion
Logs are generated whenever a specific OVN ACL triggers an event that meets the action criteria set in your logging annotation. In the example, an event in which any of the namespaces with the label `tenant: red` are denied egress to any of the namespaces with the label `tenant: orange`, a log is generated.
```
The 2nd sentence is confusing to me. If you are referring to the code sample above, then it seems incorrect because the orange tenant has an Allow action:

```
name: "allow-ns-tenant-2"
      action: "Allow"
      to:
      - namespaces:
           matchLabels:
             tenant: orange
```
if I'm totally misunderstanding, then I'd consider rephrasing/rewording the sentence for clarity.

Author: Good catch thank you.

2024-05-30T13:37:37.101Z|00034|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Ingress:2", verdict=pass, severity=warning, direction=to-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:80:02:14,dl_dst=0a:58:0a:80:02:11,nw_src=10.128.2.20,nw_dst=10.128.2.17,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=46142,tp_dst=8080,tcp_flags=ack
2024-05-30T13:37:37.101Z|00035|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Ingress:2", verdict=pass, severity=warning, direction=to-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:80:02:14,dl_dst=0a:58:0a:80:02:11,nw_src=10.128.2.20,nw_dst=10.128.2.17,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=46142,tp_dst=8080,tcp_flags=fin|ack
2024-05-30T13:37:37.101Z|00036|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Ingress:2", verdict=pass, severity=warning, direction=to-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:80:02:14,dl_dst=0a:58:0a:80:02:11,nw_src=10.128.2.20,nw_dst=10.128.2.17,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=46142,tp_dst=8080,tcp_flags=ack
----
====

> The following table describes ANP annotation values:

Reviewer 2: Suggestion: are you expecting there to be more values added to this table over time? If not, it might be simpler to remove the table and just say "the following values are supported for the `k8s.ovn.org/acl-logging` annotation" or something similar.

// Module included in the following assemblies:
//
// * networking/openshift_network_security/AdminNetworkPolicy/logging-anp-policy.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-audit-logging_{context}"]
= AdminNetworkPolicy audit logging

Audit logging is enabled per `AdminNetworkPolicy` CR by annotating an ANP policy with the `k8s.ovn.org/acl-logging` key such as in the following example:

.Example of annotation for `AdminNetworkPolicy` CR
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  annotations:
    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "alert", "pass" : "warning" }'
  name: anp-tenant-log
spec:
  priority: 5
  subject:
    namespaces:
      matchLabels:
        tenant: backend-storage

>   ingress:

Reviewer 3: My mistake. 

ingress - Allow, Allow, Pass
egress- Allow, Deny, Pass
@JoeAldinger Could you fix the action in this rule to Pass.
<pre>
- name: "pass-all-ingress-prod-sec"
      action: "Allow"
      from:
      - namespaces:
          matchLabels:
              tenant: prod-sec
</pre>

>   ingress:

Reviewer 3: Log message themselves do not change as they use policy name which is unchanged.

Now we have 
Allow that has ingress and Egress
Pass that has ingress and Egress
Deny is only Egress

>   ingress:

Reviewer 3: ok, we should be good on log messages matching the ANP exactly.

Allow verdict - 2 Ingress and 1 Egress
Pass verdict: 1 Egress and 1 ingress
Deny verdict - 1 Egress

>   ingress:

Author: @tssurya updated the logs. Thank you, @asood-rh

An example log entry resembles the following:

.Example ACL log entry for `Deny` action of the `AdminNetworkPolicy` named `anp-tenant-log`
[%collapsible]
====
[source,text]
----
2024-05-30T13:28:04.924Z|00007|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Egress:0", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:80:02:11,dl_dst=0a:58:0a:80:02:12,nw_src=10.128.2.17,nw_dst=10.128.2.18,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=35806,tp_dst=8080,tcp_flags=syn
2024-05-30T13:28:05.962Z|00008|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Egress:0", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:80:02:11,dl_dst=0a:58:0a:80:02:12,nw_src=10.128.2.17,nw_dst=10.128.2.18,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=35806,tp_dst=8080,tcp_flags=syn
2024-05-30T13:28:08.010Z|00009|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Egress:0", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:80:02:11,dl_dst=0a:58:0a:80:02:12,nw_src=10.128.2.17,nw_dst=10.128.2.18,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=35806,tp_dst=8080,tcp_flags=syn

> 2024-05-30T13:31:54.173Z|00010|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Ingress:0", verdict=drop, severity=alert, direction=to-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:80:02:12,dl_dst=0a:58:0a:80:02:11,nw_src=10.128.2.18,nw_dst=10.128.2.17,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=32834,tp_dst=8080,tcp_flags=syn

Reviewer 3: @JoeAldinger We can remove all the Ingress  entries for Deny action.

Reviewer 3: line number 85, 86 and 87.

// Module included in the following assemblies:
//
// * networking/openshift_network_security/AdminNetworkPolicy/logging-anp-policy.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-audit-logging_{context}"]
= AdminNetworkPolicy audit logging

Audit logging is enabled per `AdminNetworkPolicy` CR by annotating an ANP policy with the `k8s.ovn.org/acl-logging` key such as in the following example:

.Example of annotation for `AdminNetworkPolicy` CR
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  annotations:
    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "alert", "pass" : "warning" }'
  name: anp-tenant-log
spec:
  priority: 5
  subject:
    namespaces:
      matchLabels:

>         tenant: backend-storage #Selects all pods owned by storage tenant

Reviewer: ```suggestion
        tenant: backend-storage # Selects all pods owned by storage tenant
```

  ingress:
    - name: "allow-all-ingress-product-development-and-customer" # Product development and customer tenant ingress to backend storage.
      action: "Allow"
      from:
      - pods:
          namespaceSelector:
            matchExpressions:
            - key: tenant
              operator: In
              values:
              - product-development
              - customer
          podSelector: {}
    - name: "pass-all-ingress-product-security"
      action: "Pass"
      from:
      - namespaces:
          matchLabels:
              tenant: product-security

>     - name: "deny-all-ingress" #Ingress to backend from all other tenants.

Reviewer: ```suggestion
    - name: "deny-all-ingress" # Ingress to backend from all other pods in the cluster.
```

      action: "Deny"
      from:
      - namespaces: {}
  egress:
    - name: "allow-all-egress-product-development"
      action: "Allow"
      to:
      - pods:
          namespaceSelector:
            matchLabels:
              tenant: product-development
          podSelector: {}
    - name: "pass-egress-product-security"
      action: "Pass"
      to:
      - namespaces:
           matchLabels:
             tenant: product-security

>     - name: "deny-all-egress" # Egress from backend denied to all other tenants.

Reviewer: ```suggestion
    - name: "deny-all-egress" # Egress from backend denied to all pods.
```

      action: "Deny"
      to:
      - namespaces: {}
----
====

> Logs are generated whenever a specific OVN ACL triggers an event that meets the action criteria set in your logging annotation. For example, an event in which any of the namespaces with the label `tenant: product-development` accesses the namespaces with the label `tenant: backend-storage`, a log is generated.

Reviewer: https://github.com/openshift/openshift-docs/pull/75310#discussion_r1620326542 this comment has not been fixed.
I still don't follow "OVN ACL triggers" terminology. PTAL.

[NOTE]
====
ACL logging is limited to 60 characters. If your ANP `name` field is long, the rest of the log will be truncated.
====

The following is a direction index for the examples log entries that follow:

[cols=".^4,.^6a",options="header"]
|====
|Direction|Rule

|Ingress
|
Rule0:: Allow from tenant `product-development` and `customer`; Ingress0: `Allow`
Rule1:: Pass from `product-security`; Ingress1: `Pass`

> Rule2::	Deny all ingress from all other tenants; Ingress2: `Deny`

Reviewer: ```suggestion
Rule2::	Deny ingress from all other pods; Ingress2: `Deny`
```

|Egress
|

> Rule0:: Allow to `product-dev`; Egress0: `Allow`

Reviewer: ```suggestion
Rule0:: Allow to `product-development`; Egress0: `Allow`
```

> Rule2::	Deny all ingress from all other pods; Ingress2: `Deny`

Reviewer: ```suggestion
Rule2::	Deny ingress from all pods; Ingress2: `Deny`
```

Reviewer: why was https://github.com/openshift/openshift-docs/pull/75310#discussion_r1635148554 not fixed?

Author: I just missed it.

Rule1:: Pass to `prod-security`; Egress1: `Pass`

> Rule2:: Deny all egress from all other tenants; Egress2: `Deny`

Reviewer: ```suggestion
Rule2:: Deny egress to all other tenants; Egress2: `Deny`
```

> Rule2:: Deny egress to all other tenants; Egress2: `Deny`

Reviewer: ```suggestion
Rule2:: Deny egress to all other pods; Egress2: `Deny`
```

// Module included in the following assemblies:
//
// * networking/openshift_network_security/AdminNetworkPolicy/logging-anp-policy.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-audit-logging_{context}"]
= AdminNetworkPolicy audit logging

Audit logging is enabled per `AdminNetworkPolicy` CR by annotating an ANP policy with the `k8s.ovn.org/acl-logging` key such as in the following example:

.Example of annotation for `AdminNetworkPolicy` CR
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  annotations:
    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "alert", "pass" : "warning" }'
  name: anp-tenant-log
spec:
  priority: 5
  subject:
    namespaces:
      matchLabels:
        kubernetes.io/metadata.name: tenant
  egress:
    - name: deny-ns-tenant-1
      action: Deny
      to:
      - namespaces:
           matchLabels:

>              kubernetes.io/metadata.name: tenant_1

Reviewer 3: Change tenant_1 to tenant-1

    - name: allow-ns-tenant-2
      action: Allow
      to:
      - namespaces:
           matchLabels:

>              kubernetes.io/metadata.name: tenant_2

Reviewer 3: tenant-2

    - name: pass-ns-tenant-3
      action: Pass
      to:
      - namespaces:
           matchLabels:

>              kubernetes.io/metadata.name: tenant_3

Reviewer 3: tenant-3

// Module included in the following assemblies:
//
// * networking/openshift_network_security/AdminNetworkPolicy/logging-anp-policy.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-audit-logging_{context}"]
= AdminNetworkPolicy audit logging

Audit logging is enabled per `AdminNetworkPolicy` CR by annotating an ANP policy with the `k8s.ovn.org/acl-logging` key such as in the following example:

.Example of annotation for `AdminNetworkPolicy` CR
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  annotations:
    k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "alert", "pass" : "warning" }'
  name: anp-tenant-log
spec:
  priority: 5
  subject:
    namespaces:
      matchLabels:

>         kubernetes.io/metadata.name: tenant

Author: All namespaces with the color red are applied here when you use `tenant: red`

>   egress:

Author: rule and example updated.

    - name: deny-ns-tenant-1
      action: Deny
      to:
      - namespaces:
           matchLabels:
             kubernetes.io/metadata.name: tenant-1
    - name: allow-ns-tenant-2
      action: Allow
      to:
      - namespaces:
           matchLabels:
             kubernetes.io/metadata.name: tenant-2
    - name: pass-ns-tenant-3
      action: Pass
      to:
      - namespaces:
           matchLabels:
             kubernetes.io/metadata.name: tenant-3
----
====

> Logs are generated whenever a specific ACL meets the criteria set in your annotation. In the example, an event in which the `tenant` namespace is denied egress to the `tenant_1` namespace a log is generated.

Reviewer 3: I believe it is type of log based on action with verdict drop,allow or pass.

> Logs are generated whenever a specific ACL meets the criteria set in your annotation. In the example, an event in which the `tenant` namespace is denied egress to the `tenant_1` namespace a log is generated.

Author: Clarified by changing ACL to action.

> The logging format is compatible with syslog as defined by RFC5424. The syslog facility is configurable and defaults to `local0`. An example log entry might resemble the following:

Author: Yes, it is. I wasn't sure if it should be included or not. I thought not duplicating meant that we were assuming the user understood something. I thought it was perhaps better to make it explicit rather than left implicit.

> The logging format is compatible with syslog as defined by RFC5424. The syslog facility is configurable and defaults to `local0`. An example log entry might resemble the following:

Reviewer: I think since its under the umbrella main section of "audit logging" it is already explicit and apparent so IMHO no need to call this out in every section.

----
2024-05-28T20:53:02.612Z|00025|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Egress:0", verdict=drop, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:83:00:0e,dl_dst=0a:58:0a:83:00:15,nw_src=10.131.0.14,nw_dst=10.131.0.21,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=59714,tp_dst=8080,tcp_flags=syn
2024-05-28T21:00:08.706Z|00033|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Egress:1", verdict=allow, severity=alert, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:83:00:0e,dl_dst=0a:58:0a:83:00:10,nw_src=10.131.0.14,nw_dst=10.131.0.16,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=60156,tp_dst=8080,tcp_flags=ack
2024-05-28T21:01:23.445Z|00039|acl_log(ovn_pinctrl0)|INFO|name="ANP:anp-tenant-log:Egress:2", verdict=pass, severity=warning, direction=from-lport: tcp,vlan_tci=0x0000,dl_src=0a:58:0a:83:00:0e,dl_dst=0a:58:0a:83:00:14,nw_src=10.131.0.14,nw_dst=10.131.0.20,nw_tos=0,nw_ecn=0,nw_ttl=64,nw_frag=no,tp_src=48320,tp_dst=8080,tcp_flags=ack
----

The following table describes ANP annotation values:

.Audit logging AdminNetworkPolicy annotation
[cols=".^4,.^6a",options="header"]
|====
|Annotation|Value

|`k8s.ovn.org/acl-logging`
|

> You must specify at least one of `allow`, `deny`, or `pass` to enable audit logging for a namespace.

Reviewer 3: I tested it manually, it worked. When automating the test, I can tweak the various combinations and check for log messages.

// Module included in the following assemblies:
//
// * networking/openshift_network_security/AdminNetworkPolicy/logging-anp-policy.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-audit-logging_{context}"]
= AdminNetworkPolicy audit logging

> The OVN-Kubernetes network plugin uses Open Virtual Network (OVN) access control list (ACL) to manage egress firewalls and network policies. Audit logging exposes `allow` and `deny` ACL events for `NetworkPolicy` and `BaselineAdminNetworkPolicy` custom resources (CR). Logging also exposes `allow`, `deny`, and `pass` ACL events for `AdminNetworkPolicy` (ANP) CR.

Reviewer 3: Could we rephrase this to indicate that before it was just egress firewalls and network policies but now baseline admin network policies and admin network policies before we elaborate on the various ACLs and their logging.

Author: This is a bit of what I am trying to get at with the importance of naming things. Here, when I speak of `network policy` I mean: `ANP, BANP, and NP`. Maybe we could say something like `egress firewall and both cluster scoped and namespace scoped network policy`?

Reviewer 3: may be  say all ( ANP, BANP and NP) network policies as (B)ANP have joined the group.

You can configure the destination for audit logs, such as a syslog server or a UNIX domain socket.
Regardless of any additional configuration, an audit log is always saved to `/var/log/ovn/acl-audit-log.log` on each OVN-Kubernetes pod in the cluster.

Audit logging is enabled per `AdminNetworkPolicy` CR by annotating an ANP policy with the `k8s.ovn.org/acl-logging` key such as in the following example:

.Example of annotation for `AdminNetworkPolicy` CR
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
 annotations:
  k8s.ovn.org/acl-logging: '{ "deny": "alert", "allow": "alert", "pass" : "warning" }'
 name: anp-tenant-log
spec:
 priority: 5
 subject:
  namespaces:
   matchLabels:
     kubernetes.io/metadata.name: tenant
 egress:
 - name: "deny-ns-tenant-1"
  action: "Deny"
  to:
  - namespaces:

>     matchLabels:

Reviewer 3: We need to fix indentation for matchLabels and the following key kubernetes.io/metadata.name

     kubernetes.io/metadata.name: tenant_1
 - name: "allow-ns-tenant-2"
  action: "Allow"
  to:
  - namespaces:

>     matchLabels:

Reviewer 3: Same as above

     kubernetes.io/metadata.name: tenant_2
 - name: "pass-ns-tenant-3"
  action: "Pass"
  to:
  - namespaces:
    matchLabels:
     kubernetes.io/metadata.name: tenant_3
----
====

> Logs are generated whenever a specific ACL meets the criteria set in your annotation. In the example, an event in which the `tenant 1` namespace is denied egress to the `tenant` namespace a log is generated.

Reviewer 3: Egress to tenant1 is denied from tenant, source is tenant and destination is tenant1

Author: whoops

> The logging format is compatible with syslog as defined by RFC5424. The syslog facility is configurable and defaults to `local0`. An example log entry might resemble the following:

Reviewer 3: Do you want to move this to where we describe policyAuditConfig:?

Author: I'm not sure I understand what you mean. I was trying to follow the structure of the [here](https://docs.openshift.com/container-platform/4.15/networking/ovn_kubernetes_network_provider/logging-network-policy.html).
